core/src/main/java/com/sequenceiq/cloudbreak/converter/stack/cluster/ServiceEndpointCollector.java

```diff
@@ -0,0 +1,189 @@
+
+    public ServiceEndpointCollector(BlueprintValidator blueprintValidator, StackServiceComponentDescriptors stackServiceComponentDescs) {
+        this.blueprintValidator = blueprintValidator;
+        this.stackServiceComponentDescs = stackServiceComponentDescs;
+    }
+
+    public Map<String, String> collectServiceUrlsForPorts(Cluster cluster, String ambariIp) {
+        Set<HostGroup> hostGroups = cluster.getHostGroups();
+        Blueprint blueprint = cluster.getBlueprint();
+
+        Map<String, String> result = new HashMap<>();
+        List<Port> ports = NetworkUtils.getAllPorts();
+        try {
+            JsonNode hostGroupsNode = blueprintValidator.getHostGroupNode(blueprint);
+            Map<String, HostGroup> hostGroupMap = blueprintValidator.createHostGroupMap(hostGroups);
+            for (JsonNode hostGroupNode : hostGroupsNode) {
+                String hostGroupName = blueprintValidator.getHostGroupName(hostGroupNode);
+                JsonNode componentsNode = blueprintValidator.getComponentsNode(hostGroupNode);
+                HostGroup actualHostgroup = hostGroupMap.get(hostGroupName);
+                final String serviceAddress;
+                if (actualHostgroup.getConstraint().getInstanceGroup() != null) {
+                    InstanceMetaData next = actualHostgroup.getConstraint().getInstanceGroup().getNotDeletedInstanceMetaDataSet().iterator().next();
+                    serviceAddress = next.getPublicIpWrapper();
+                } else {
+                    serviceAddress = actualHostgroup.getHostMetadata().iterator().next().getHostName();
+                }
+                for (JsonNode componentNode : componentsNode) {
+                    String componentName = componentNode.get("name").asText();
+                    StackServiceComponentDescriptor componentDescriptor = stackServiceComponentDescs.get(componentName);
+                    Map<String, String> collectedPorts = collectServicePorts(ports, ambariIp, serviceAddress, componentDescriptor, cluster);
+                    result.putAll(collectedPorts);
+                }
+            }
+        } catch (Exception ignored) {
+            return result;
+        }
+        return result;
+    }
+
+    public String getAmbariServerUrl(Cluster cluster, String ambariIp) {
+        String orchestrator = cluster.getStack().getOrchestrator().getType();
+        if (ambariIp != null) {
+            Gateway gateway = cluster.getGateway();
+            if (YARN.equals(orchestrator)) {
+                return String.format("http://%s:8080", ambariIp);
+            } else {
+                if (gateway != null) {
+                    Optional<GatewayTopology> gatewayTopologyWithAmbari = getGatewayTopologyWithAmbari(gateway);
+                    return gatewayTopologyWithAmbari
+                            .map(gt -> getAmbariUrlFromGatewayTopology(ambariIp, gateway, gt))
+                            .orElse(String.format("https://%s/ambari/", ambariIp));
```

Comment:
Return the same value from two places, code duplication issue or control flow issue

Reply:
ambari has the same address in these two different cases:
1) the user did not expose ambari in any of the topologies
2) there is no gateway definition in the request at all

Reply:
although I still have to ask if it's the expected behaviour